src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/PackageReferenceProject.cs

```diff
@@ -150,83 +185,101 @@ public async Task<ProjectPackages> GetInstalledAndTransitivePackagesAsync(bool i
-                    });
-
-                lock (_transitiveOriginsLock)
+                else
                 {
-                    TransitiveOriginsCache = transitiveOrigins;
+                    // Get Transitive packages without Transitive Origins
+                    transitivePackagesWithOrigins = calculatedTransitivePackages
+                        .Select(packageRef => new TransitivePackageReference(packageRef));
                 }
             }
-            else
-            {
-                // Get Transitive packages without Transitive Origins
-                transitivePackagesWithOrigins = calculatedTransitivePackages
-                    .Select(packageRef => new TransitivePackageReference(packageRef));
-            }
 
             List<TransitivePackageReference> transitivePkgsResult = transitivePackagesWithOrigins.ToList(); // Materialize results before setting IsInstalledAndTransitiveComputationNeeded flag to false
-            IsInstalledAndTransitiveComputationNeeded = false;
 
             // Refresh cache
             lock (_installedAndTransitivePackagesLock)
             {
                 InstalledPackages = installedPackages;
-                TransitivePackages = transitivePackages;
+            }
+            if (includeTransitivePackages)
+            {
+                lock (_installedAndTransitivePackagesLock)
+                {
+                    TransitivePackages = transitivePackages;
+                }
             }
 
+            IsInstalledAndTransitiveComputationNeeded = false;
+
             return new ProjectPackages(calculatedInstalledPackages, transitivePkgsResult);
         }
 
         /// <inheritdoc/>
-        public virtual async Task<ProjectPackages> GetInstalledAndTransitivePackagesAsync(CancellationToken token) => await GetInstalledAndTransitivePackagesAsync(includeTransitiveOrigins: false, token);
+        public virtual async Task<ProjectPackages> GetInstalledAndTransitivePackagesAsync(CancellationToken token) => await GetInstalledAndTransitivePackagesAsync(includeTransitivePackages: true, includeTransitiveOrigins: false, token);
 
-        protected abstract IEnumerable<PackageReference> ResolvedInstalledPackagesList(IEnumerable<LibraryDependency> libraries, NuGetFramework targetFramework, IReadOnlyList<LockFileTarget> targets, T installedPackages);
+        protected abstract IEnumerable<PackageReference> ResolvedInstalledPackagesList(IEnumerable<LibraryDependency> libraries, NuGetFramework targetFramework, IList<LockFileTarget> targets, T installedPackages);
 
-        protected abstract IReadOnlyList<PackageReference> ResolvedTransitivePackagesList(NuGetFramework targetFramework, IReadOnlyList<LockFileTarget> targets, T installedPackages, T transitivePackages);
+        protected abstract IReadOnlyList<PackageReference> ResolvedTransitivePackagesList(NuGetFramework targetFramework, IList<LockFileTarget> targets, T installedPackages, T transitivePackages);
```

Comment:
I don't understand why the `targets` type was changed from `IReadOnly<T>` to `<IList<T>`. I don't see where it's being mutated.

Haven't we already had an instance of `PackageReferenceProject.GetInstalledPackageAsync()` not being thread-safe?  Does this change add risk that the implementation might become non-thread safe again?

Reply:
See changes in PackageReferenceProject.cs

![image](https://user-images.githubusercontent.com/1192347/204672776-d2c51a52-4ca1-4e30-8427-0d6ea81c4fde.png)

IList has methods to mutate the list, whereas `IReadOnly...<T>` not, and I want to avoid type conversions to avoid performance issues.

Reply:
Regarding thread-safety, this PR calls GetInstalledAndTransitivePackageAsync(), which is a method that is considered thread-safe. 

There are tests that fall into this scenario, see 

https://github.com/NuGet/NuGet.Client/blob/c0371f7165bea201aded1a8013ca0b17e2ad0092/test/NuGet.Clients.Tests/NuGet.PackageManagement.VisualStudio.Test/CpsPackageReferenceProjectTests.cs#LL3908C147-L3908C147


I can add a custom test for this API, if you want.

Reply:
Also, there's a test in this PR to test GetInstalledPackagesAsync With GetInstalledAndTransitivePackagesAsync calls at the same time.

GetInstalledPackagesAsync_WithGetInstalledAndTransitivePackagesCallAtSameTime_DoesNotDeadlockAsync

Reply:
Added new test: GetInstalledPackagesAsync_WithMultipleCalls_DoesNotDeadlockAsync